doc/proposals/named-networks.md

```diff
@@ -0,0 +1,231 @@
+# Support for named networks
+
+# Table of contents
+
+- [Introduction](#introduction)
+  - [Goal](#goal-of-this-proposal)
+- [Design](#design)
+  - [Changes in IPAM Config](#changes-in-ipam-config)
+  - [Changes in Modules](#changes-in-modules)
+- [Summary](#summary)
+- [Discussions and Decisions](#discussions-and-decisions)
+
+<hr>
+
+## Introduction
+
+When whereabouts assigns an IP to a Pod this fact is recorded in a CR of kind `IPPool` that has its name derived from the CIDR range in question.
+This CR is stored in the namespace of whereabouts (`kube-system` by default).
+
+Should the user configure multiple overlapping ranges, it is possible to configure whereabouts to allow assigning duplicate IPs.
+
+However, since the storage of the assignments is done in a CR that is named like the CIDR range, it is not possible to configure *the same CIDR range* twice and have whereabouts assign from the ranges independently.
+
+This is, for example, useful in multi-tenant situations where more than one group is responsible for selecting CIDR ranges.
+
+### Goal of this Proposal
+
+- Allow configuring the same CIDR range multiple times (e.g. in separate multus-`NetworkAttachmentDefinition`s)
+
+<hr>
+
+## Design
+
+The network configuration already has a field `name`:
+
+```json
+{
+      "cniVersion": "0.3.0",
+      "name": "whereaboutsexample",
```

Comment:
I think there's some complexity with "name" that it might make users have unexpected division of pools of IPs for existing net-attach-defs/CNI configs.

Given:

```
{
	"cniVersion": "0.3.0",
	"name": "macvlan-red",
	"type": "macvlan",
	"master": "eth0",
	"mode": "bridge",
	"ipam": {
		"type": "whereabouts",
		"range": "192.168.2.225/28"
	}
}

{
	"cniVersion": "0.3.0",
	"name": "macvlan-blue",
	"type": "macvlan",
	"master": "eth1",
	"mode": "bridge",
	"ipam": {
		"type": "whereabouts",
		"range": "192.168.2.225/28"
	}
}
```

Users would expect  `192.168.2.225/28` to be taken from the same IP pool across two configurations, the way it exists today.

Reusing the name would cause users to have to change their net-attach-defs.

If we add a new "name" type field to the whereabouts configuration in the IPAM section, then users would have to specifically add it, and... If it's not provided we could draw from a "default" pool. Such as:

```
{
	"cniVersion": "0.3.0",
	"name": "macvlan-color",
	"type": "macvlan",
	"master": "eth1",
	"mode": "bridge",
	"ipam": {
	    "pool_name": "whereabouts-blue-pool",
		"type": "whereabouts",
		"range": "192.168.2.225/28"
	}
}
```

Reply:
Maybe we need kind of a combination of both... Like, ideally not having the generated name is overall better.

But, I kind of don't want to break behavior that people might expect with existing CRDs.

Maybe if no name is specified, then we generate it. But, we honor the name when used. 

I'm still thinking it through -- Ayush also brought up that it's possible this could also cause confusion for users who forget to name a range, and then those pools get divided.

Reply:
How about implementing a comprehensive comparator for IP pools, which works on a combination of `pool_name` and `range` ?

If the user does not define the `pool_name`, we can add a wildcard for it. In the cases when the _wildcard_exists, we only use the `range` for comparison.

Reply:
Instead of using `name` we can also be more explicit and call it something like `addressCollisionDomain` with a default of `""` -- that would also make the code more readable as the names of the `IPPools` would "just" be `<CIDR>-<addressCollisionDomain>`.